[Product Multi-selection] Unselecting and removing products from Order

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8922

## Description
This PR deals with the "unselect" logic for Products and Product Variations while editing an Order. The same way we add a product to an Order when selecting it, we also want to remove a product from an Order if we unselect it.

## Changes
- Renamed `addProductViewModel` to `handleProductsViewModel`, as the method does add and/or remove products as needed. 
- This `handleProductsViewModel` is no longer lazy, as needs to be to re-initialized and pass the updated selected products across the `EditableOrderViewModel` and `ProductSelectorViewModel`.
- Added `selectedProducts` and `selectedProductVariations` published property wrappers, in order to track which items are selected/unselected. 
- We use the existing `RemoteOrderSynchronizer.setProduct()` to add or update the products and variations

## Testing instructions

- Go to Menu > Settings > Experimental features > Switch the `Product Multi-Selection` feature on.

### Flow 1
- Go to Orders > `+` > `+ Add Product`
- Select a Product > Tap on Done > See how is added to the Order
- Tap again on `+ Add Product`
- See how the product toggle appears as selected
- Tap on the selected product to unselect it > Tap on Done > See how is removed from the Order
- The same behaviour should work for Product Variations

### Flow 2
- Go to Orders > `+` > `+ Add Product`
- Select a Product > Tap on Done > See how is added to the Order
- Tap on the `-` (minus) icon to remove the product from the Order > Confirm by tapping `Remove Product from Order``
- Tap on `+ Add Product` > See how the Product toggle appears as unselected
- The same behaviour should work for Product Variations

iamgabrielma committed:
remove trailing closure to make arguments explicit

iamgabrielma committed:
Add/remove product to Order on select/unselect state

iamgabrielma committed:
Add/remove product variation to Order on select/unselect

iamgabrielma committed:
Adapt Unit Tests to feature flag usage

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -168,13 +168,18 @@ final class EditableOrderViewModel: ObservableObject {
     /// View model for the product list
     ///
     lazy var addProductViewModel = {
```
This can be changed to `handleProductsViewModel` instead.

iamgabrielma committed:
Internal comments as TODO items

iamgabrielma committed:
Remove debug prints. Add DDLogs

iamgabrielma committed:
Rename view model for addProduct to handleProducts

iamgabrielma committed:
Add method description

iamgabrielma committed:
Extract addProduct & removeProduct to methods

iamgabrielma committed:
Extract method for add & remove product variations

iamgabrielma committed:
internal comments

iamgabrielma committed:
remove internal comment

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -669,6 +683,78 @@ private extension EditableOrderViewModel {
+    /// Handles a product by adding it to the Order if selected, or removing it from the Order if unselected
+    ///
+    func handleProduct(_ product: Product) {
+        // All products
+        if !allProducts.contains(product) {
+            allProducts.append(product)
+        }
+
+        if !selectedProducts.contains(product) {
+            // Add Product to the Order
+            selectedProducts.append(product)
+            addProductToOrder(product)
+        } else {
+            // Remove Product from the Order
+            selectedProducts.removeAll(where: { $0?.productID == product.productID })
+            removeProductFromOrder(product)
+        }
+    }
+
+    /// Handles a product variation by adding it to the Order if selected, or removing it from the Order if unselected
+    ///
+    func handleProductVariation(_ variation: ProductVariation, parent product: Product) {
+        if !allProducts.contains(product) {
+            allProducts.append(product)
+        }
+
+        if !allProductVariations.contains(variation) {
+            allProductVariations.append(variation)
+        }
+
+        if !selectedProductVariations.contains(variation) {
+            // Add Product variation to the Order
+            selectedProductVariations.append(variation)
+            addProductVariationToOrder(variation)
+        } else {
+            // Remove Product variation from the Order
+            selectedProductVariations.removeAll(where: { $0?.productVariationID == variation.productVariationID })
+            removeProductVariationFromOrder(variation)
+        }
+    }
```
As there's certain logic repetition, would you say is worth it to attempt and make an unique method here for both `handleProduct` and `handleProductVariation`? Or to keep these separate for readability?

iamgabrielma committed:
Merge branch 'issue/8920-resolve-multiselection-toggle-state' into issue/8922-remove-product-on-unselect

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift

iamgabrielma committed:
Merge branch 'trunk' into issue/8922-remove-product-on-unselect

iamgabrielma committed:
Use handleProduct method in onProductSelected

iamgabrielma committed:
Add productMultiSelectionM1 to MockFeatureFlagService

iamgabrielma committed:
Inject ProductSelectorViewModel’s FeatureFlag from init

iamgabrielma committed:
Adapt incrementQuantity test to multiple selection

iamgabrielma committed:
Remove unnecessary check

Ensuring that product and variation lists are not cleared after resync is already handled within `addProductToOrder(_:)`

iamgabrielma committed:
Pass salectedItemIDs to ProductSelectorViewModel

In order to refresh the list of selectedItemIDs each time we open the product selector, we need to re-initialize the ProductSelectorViewModel, so it can’t be lazy. This is necessary to keep the state of selected products if these are removed directly from the Order, rather than from the product selector.

iamgabrielma committed:
Add Unit Tests for ProductSelectorViewModel

iamgabrielma committed:
Update models to account for Product Variations

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -194,6 +195,20 @@ final class EditableOrderViewModel: ObservableObject {
+
+    /// Keeps track of the total of selected Products and Product Variations, if any
+    ///
+    var selectedItems: [Int64] {
```
I would rename this variable to make it more meaningful `selectedItemsProductIDs`, and update the comment accordingly

iamgabrielma replied:
Added some Unit Tests here https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/eca144ab112a22f61c7dfd70f184b51fc4a97b45 and renamed to `selectedProductsAndVariationsIDs` here: https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/08f84066f709be255b272fe6938f6e23770004c0

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -167,23 +167,24 @@ final class EditableOrderViewModel: ObservableObject {
     /// View model for the product list
     ///
-    lazy var addProductViewModel = {
+    var handleProductsViewModel: ProductSelectorViewModel {
```
In the [OrderForm](https://github.com/woocommerce/woocommerce-ios/blob/0236d06f37493a0b4ca75719eeec7f0a74b9248d/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/OrderForm.swift#L285) class it expects it to be lazy:

`viewModel.addProductViewModel.clearSearchAndFilters()`

Did you take that into account?

Furthermore, with it being a property, I would expect it as well that it was created once, without being generated every time. If we want to do that, I would create a function to make it clear:

`func generateProductSelectorViewModel()`

From [here](https://stackoverflow.com/questions/24035276/computed-read-only-property-vs-function-in-swift):

> It seems to me that it's mostly a matter of style: I strongly prefer using properties for just that: properties; meaning simple values that you can get and/or set. I use functions (or methods) when actual work is being done

iamgabrielma replied:
I believe I need some guidance on how to move forward with this bit:

For creating the `ProductSelectorViewModel` I experimented with 3 options: 
1. Lazy ( currently in trunk)
2. Computed (changes in this PR)
3. Method call

And only the 2nd option ( computed ) seems to work as we want. The problem is within the `Flow 2` explained above in the `Testing Instructions`:
- Go to Orders > `+` > `+ Add Product`
- Select a Product > Tap on Done > See how is added to the Order
- Tap on the `-` (minus) icon to remove the product from the Order > Confirm by tapping `Remove Product from Order`
- Tap on + Add Product > See how the Product toggle appears as unselected

Only making the view model computed we seem to update the selected products properly, the other two will keep the state and products will appear selected despite had been removed from the Order.

From what I understand, if the view model being lazy we're initializing this once with whatever values we have on `selectedItemIDs`, so values are not updated as needed.

The same seems to happen with the 3rd option by using a function call, which was unexpected as I would expect that this is re-instantiated with the new values. Could have something to do with the view model being an `@ObservableObject`? I attempted to add a `deinit` breakpoint to see when it's reinstantiated but this doesn't seem to be triggered, so that could be why both non-computed approaches show the same issue.

Perhaps there's a way to not compute the whole view model, but just what's needed?

iamgabrielma replied:
> Method call

I found the way to make the method call approach work by also changing the [ProductSelectorView.Configuration](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/OrderForm.swift#L354) initializer into an static method, so we keep the model updated when the view calls it. However, this causes other UI issues that need to be resolved, so I'm switching the PR back to `draft` for the moment.

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -321,6 +336,7 @@ final class EditableOrderViewModel: ObservableObject {
     func createProductRowViewModel(for item: OrderItem, canChangeQuantity: Bool) -> ProductRowViewModel? {
         guard item.quantity > 0, // Don't render any item with `.zero` quantity.
               let product = allProducts.first(where: { $0.productID == item.productID }) else {
+            selectedProducts.removeAll(where: { $0?.productID == item.productID })
```
I am wondering if this is the right place for doing this. After all, on this function we just want to create a `ProductRowViewModel` instance and return it, removing that item from the selected products is a side effect that is not expected from the function name.

iamgabrielma replied:
There seems to be better places indeed! I moved the method to `removeItemFromOrder(_:)` instead,  here: https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/1e4d574bb5b4bd58be4a3f0c09c5c6398566c222 , as is where we specifically deal with synching, updating, and tracking when an item is removed from an Order.

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -677,6 +693,61 @@ private extension EditableOrderViewModel {
+
+    /// Handles a product by adding it to the Order if selected, or removing it from the Order if unselected
+    ///
+    func handleProduct(_ product: Product) {
```
I would rename this function to specify what's doing with the product, `handleProduct` sounds too generic for me and I wouldn't know what's happening inside.

iamgabrielma replied:
Renamed to `addOrRemoveProductFromOrder` here: https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/cb86a9852a0da3321f719b4a40662a1577e60a8d

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -677,6 +693,61 @@ private extension EditableOrderViewModel {
+
+    /// Handles a product variation by adding it to the Order if selected, or removing it from the Order if unselected
+    ///
+    func handleProductVariation(_ variation: ProductVariation, parent product: Product) {
```
Same here about function naming.

iamgabrielma replied:
Renamed to `addOrRemoveProductVariationFromOrder` here https://github.com/woocommerce/woocommerce-ios/commit/cb86a9852a0da3321f719b4a40662a1577e60a8d

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift:
```diff
@@ -280,9 +280,9 @@ private struct ProductsSection: View {
+                                    viewModel: viewModel.handleProductsViewModel)
                         .onDisappear {
-                            viewModel.addProductViewModel.clearSearchAndFilters()
+                            viewModel.handleProductsViewModel.clearSearchAndFilters()
```
As explained before, I am wondering if this has any effect now that we are creating a new view model any time we call the property.

iamgabrielma replied:
I removed the method call here: https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/0be8ea5de00277c7a460532828643df66498dcb6 and added further tests.

Calling this method `.onDisappear` was necessary for clearing the search terms  when the `ProductSelectorViewModel` was lazily initialized, however now that we re-instantiate the object, the search terms are cleared by default. 

Within the same closure there's also an (unrelated to this feature) assignment `navigationButtonID = UUID()` to re-instantiate the button ID as needed, which we accomplish now automatically when the object is re-instantiated, so we shouldn't need this either, but I left it outside of the PR and will make a note to come back to it.

jostnes commented:
Sorry I missed looking at the last few PRs; using this build, I found some issues that could have happened in previous PRs, but sharing them here in case issues are not known (I tried to look for them but couldn't find the issues reported/mentioned in PRs on the [project board](https://github.com/woocommerce/woocommerce-ios/projects/54)):

1. For variable products, we can't select them from the Product screen even though it looks like it's clickable. Should it be made clearer for merchants that for a variable product, we need to select the product individually?

<div align="center">
  <video src="https://user-images.githubusercontent.com/17252150/223348472-13a26bfb-97ca-4bb1-b22c-6a3af9f75524.MP4"/>
</div>

2. Number displayed on the Done button is incorrect if a variable product is selected (it adds 2 to the count), but the count is corrected after clicking on Done and clicking on Add Product again (though the count is corrected, the UX is incorrect as the variable product is no longer displayed as selected):

<div align="center">
  <video src="https://user-images.githubusercontent.com/17252150/223348958-0267f8bc-709c-483d-8db6-db6fdd418cc3.MP4"/>
</div>

3. When selecting Add Products from Edit Order flow, previously selected products are not shown as selected. Also adding an already selected product will create a new row for that product (this is existing product, just to confirm that this behavior won't change as part of this project?)

<div align="center">
  <video src="https://user-images.githubusercontent.com/17252150/223349551-32ccd6ef-95f6-4817-af1d-fe23f9616ab6.MP4"/>
</div>

4. After selecting a product from search results, it clears the search results. It seems like if a merchant wants to select 2 products with a similar name, they would have to search twice in this case.

<div align="center">
  <video src="https://user-images.githubusercontent.com/17252150/223350842-462ad873-6296-4874-8780-356074ac9ef6.MP4"/>
</div>

5. There are also performance issues with product selection (an obvious delay between selections) but I understand that this will be tackled in M1.1.

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -194,6 +195,20 @@ final class EditableOrderViewModel: ObservableObject {
 
+    /// Keeps track of selected/unselected Products, if any
+    ///
+    @Published var selectedProducts: [Product?] = []
```
Are these properties necessary, or can they be taken from the `addProductViewModel?

iamgabrielma replied:
That's a good question 🤔 , at the moment these are necessary as in the `addProductViewModel` we just keep track of IDs, and we need to pass the model objects to add or remove `Products` and `ProductVariations`, but I'll reevaluate based on the direction we take with how we generate the `ProductSelectorViewModel`

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -167,23 +167,24 @@ final class EditableOrderViewModel: ObservableObject {
     /// View model for the product list
     ///
-    lazy var addProductViewModel = {
+    var handleProductsViewModel: ProductSelectorViewModel {
```
Can we add a more specific name than `handleProductsViewModel`?

iamgabrielma replied:
Definitely! Now that not only adds products, it seemed appropriate to just name it as its definition, so I renamed it to `productSelectorViewModel` here https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/aed18f77fe9bcc47cb0e40f8d42a76ba2a012e4a

iamgabrielma committed:
Rename allSelectedItemIDs. Add Unit Tests.

iamgabrielma committed:
Another allSelectedItemIDs rename

iamgabrielma committed:
Remove clearSearchAndFilters call when onDisappear

This method call is no longer needed since the view model is not lazy, and we do not retain those properties state (Additionally, the view model does not have Filters property, but does have Search).

iamgabrielma committed:
Move updating selectedProducts to removeItemFromOrder

iamgabrielma committed:
Rename handleProduct, and handleProductVariation

iamgabrielma committed:
Rename EditableOrder productSelectorViewModel

iamgabrielma commented:
Thanks for the review @toupper , this is ready for re-review when you have the time 🙇 . I'm specially interested on the [comment here](https://github.com/woocommerce/woocommerce-ios/pull/8943#discussion_r1129030624) regarding the best way on how to initialize the `ProductSelectorViewModel`.

Thanks for testing this as well @jostnes ! 💯 
> For variable products, we can't select them from the Product screen even though it looks like it's clickable. Should it be made clearer for merchants that for a variable product, we need to select the product individually?

This is expected, as we do not want them to select all variations but open the product instead and select specific ones. You're right that UI-wise could be more clear. Opened issue here:  https://github.com/woocommerce/woocommerce-ios/issues/9074

> Number displayed on the Done button is incorrect if a variable product is selected

I couldn't replicate this on my end so far, but I'm wondering if counts a single variation as product + variation at certain point. Will test further!

> When selecting Add Products from Edit Order flow, previously selected products are not shown as selected. 

Does it happen the first time you create an Order? Or only if you go through the Order editing flow for a second time (once the order has already been created, then you add more to it). In this case the products won't appear as selected indeed, so is not the best flow. I'll confirm with Android on how are we going to deal with this and the expected behaviour 👍 

> Also adding an already selected product will create a new row for that product (this is existing product, just to confirm that this behavior won't change as part of this project?)

Good question, this is how currently works for single-selection (adding a product twice will create 2 rows), so I haven't planned to change it for multi-selection. I've opened an issue here https://github.com/woocommerce/woocommerce-ios/issues/9075 to keep track of exploring possibilities. Answering your question, for the moment I'd say the behaviour will not change as part of this project.

> After selecting a product from search results, it clears the search results. It seems like if a merchant wants to select 2 products with a similar name, they would have to search twice in this case.

Good point! The current default is to clear both filters and search results once a product has been added. I've updated this existing issue: https://github.com/woocommerce/woocommerce-ios/issues/8998 to tackle search as well.

> There are also performance issues with product selection (an obvious delay between selections) but I understand that this will be tackled in M1.1.

That's right, the next iteration should tackle sync the Order only when we create it, rather than after each Product addition or removal, so should feel faster :D

toupper commented:
@iamgabrielma I tested both flows. Flow 2 tests well, however, the Flow 1 had some issues for me:

- Select a Product > Tap on Done > See how is added to the Order ❓ There is no Done button, just 1 Product Selected.
- Tap on the selected product to unselect it > Tap on Done > See how is removed from the Order ❌ 

When tapping on a selected product it's not deselected, and the product is added again to the order. (as a different product, it doesn't update the quantity):


https://user-images.githubusercontent.com/1864060/223742170-5d12c736-a2e6-4a2c-a16a-cdb7dbadc8b2.MP4

toupper commented:
> Switch the productMultiSelectionM1 feature flag to true

@iamgabrielma  I think it would be easier if we add a toggle in Experimental Feature to disable/enable this feature, as we did for others (remember to keep it only in Debug). That way we don't have to compile the project again.

<img src="https://user-images.githubusercontent.com/1864060/223742951-26f27e7e-3e53-4bcf-b4d0-627e18a54f37.jpeg" width="375">

iamgabrielma committed:
Move ProductSelectorViewModel into a generate method

joshheald commented:
@iamgabrielma As requested, I've checked Flow 1 and see the same as @toupper 

> Flow 1
> 
> Go to Orders > + > + Add Product
> Select a Product > Tap on Done > See how is added to the Order
> Tap again on + Add Product
> See how the product toggle appears as selected
> Tap on the selected product to unselect it > Tap on Done > See how is removed from the Order

The first tap on a selected product at this step puts the button into a loading state, but leaves the product selected. As @toupper found, tapping the button (not Done, it doesn't change to that) adds a duplicate of that product to the order.

Tapping the selected product more than once does deselect it and change the button title, but still doesn't remove the item from the order when you tap Done.

> The same behaviour should work for Product Variations

It doesn't... and Product variations didn't show as selected for me.

See video

https://user-images.githubusercontent.com/2472348/224736721-792ff771-9415-4095-971a-10c59d0b94fa.mp4

iamgabrielma committed:
Merge branch 'trunk' into issue/8922-remove-product-on-unselect

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift
#	WooCommerce/WooCommerceTests/Mocks/MockFeatureFlagService.swift
#	WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift

iamgabrielma committed:
Revert ProductSelector vm to lazy. Make tests compile

iamgabrielma committed:
Fix faulty productID check

iamgabrielma committed:
fix selectedProductVariations mapping wrong values

iamgabrielma committed:
remove print statement

iamgabrielma commented:
Thanks for the review @toupper and @joshheald ! This is ready for a new round 🙇 

> When tapping on a selected product it's not deselected

There were some faulty checks that are now fixed with https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/3365bdbd1add71f764ec3f074837b92c0561d707 and https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/20171f8f65e36d41da9a1dddca3ae015f0367569 . Selection and deselection state should work correctly at the moment. This also resolves the issue where product count was not correct when there were variations selected

> I think it would be easier if we add a toggle in Experimental Feature to disable/enable this feature, as we did for others (remember to keep it only in Debug). That way we don't have to compile the project again

Added via https://github.com/woocommerce/woocommerce-ios/pull/9097 , I updated this branch with trunk so you can switch back and forth now as needed.

> and the product is added again to the order. (as a different product, it doesn't update the quantity):

Yes, @jostnes reported the same issue: This is how currently works for single-selection as well (adding a product twice will create 2 rows, rather than adding +1), so I haven't targeted making this different for the moment. I've opened #9075 to keep track of exploring possibilities.

iamgabrielma committed:
toggle showing variations list on checkbox tap

toupper commented:
@iamgabrielma Thanks for the fixes. I re-tested it briefly and while I couldn't reproduce the previous issue, I found another two:

- I don't know the exact steps, but I found a state where the products in the Order Detail and Selection screens are not synchronized:

https://user-images.githubusercontent.com/1864060/225288317-d07b76f1-6587-4e27-8d16-a902ee02598b.MP4


- Removing variable products in the Order Detail screen is not synchronized with the Selection screen:

https://user-images.githubusercontent.com/1864060/225288882-8f10ef6c-1081-4894-8961-c4b3026f90eb.MP4

iamgabrielma committed:
Sync variable products when removed from Order view

iamgabrielma commented:
> Removing variable products in the Order Detail screen is not synchronized with the Selection screen:

Missed this one! Fixed here: https://github.com/woocommerce/woocommerce-ios/pull/8943/commits/312a6f4a8a52ee29b8a15707fbc78b8ac9d96774

> I don't know the exact steps, but I found a state where the products in the Order Detail and Selection screens are not synchronized:

I couldn't reproduce this one, but by the error notice displayed on screen seems to happen if the [remote Order can't be synchronized](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/EditableOrderViewModel.swift#L701), so we fire the error to let the merchant know it isn't sync. It would seem the Order is successfully created though, as otherwise we would fire [a different error](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/EditableOrderViewModel.swift#L426).

I've opened a new issue to decide how to deal with these here: https://github.com/woocommerce/woocommerce-ios/issues/9159 , for example perhaps we should dismiss the Order entirely so the merchant can redo it, or offer a retry option.

joshheald commented:
@iamgabrielma thanks for the updates, I can see that you've fixed several issues that were previously there, particularly the `.contains(product)` check that was causing problems. 

(Sorry to be the bearer of bad news, but...) I've found a deselection issue with repro steps:

The first run-through in this video shows that you can create an order, add two products to it, then go in and deselect the second product, and everything in the selector and the order is correct and in sync.

The second order created in the video shows an issue that happens if the user is less decisive, and goes in to deselect an item, then reselects it, and deselects it again before tapping the button to go back to the order. In this case, the deselected item remains on the order (incorrectly), but is removed from the selection picker. Adding it back in the selection picker adds the same item to the order again.

https://user-images.githubusercontent.com/2472348/225611728-e230411f-ee9d-4b1e-82d3-df1d6a170fe0.mp4

Repro:

1. Create an order
2. Tap `Add Products`, and select two products
3. Tap `2 products selected`
4. Tap `Add Products`, then deselect, select, and deselect again one of the products
5. Tap `1 product selected`
6. Observe that the second product is still shown on the order (incorrectly)
7. Tap `Add Products` and observe that the second product is not selected in the selection screen
8. Select that same product again, tap `2 products selected`
9. Observe that the second product is listed twice on the order

Note that the button on the selection screen behaves differently for selection and deselection. When selecting a product, it's briefly disabled while the network request to update the order is running... when deselecting a product, it doesn't get disabled. Why is that? Perhaps that's part of the cause?

I think the shared state between the selection screen and the order, and keeping the order synced with the server for every action on the selection screen is probably causing more trouble than it's worth here... it may be better to move towards a single network request when the selection screen is dismissed, providing all the updates in a batch.

iamgabrielma committed:
Merge pull request #9144 from woocommerce/issue/9120-redirect-to-variations-on-toggle-tap

[Product Multi-selection] Redirect to variations view on checkbox toggle tap

iamgabrielma commented:
Thanks for the review and the thoughtful testing @joshheald 😄 

> I think the shared state between the selection screen and the order, and keeping the order synced with the server for every action on the selection screen is probably causing more trouble than it's worth here... it may be better to move towards a single network request when the selection screen is dismissed, providing all the updates in a batch.

That seems to be the case indeed, after some discussions with @toupper as well seems that the best course of action is to move the project to a single network request that contains all necessary changes. I'll close this issue for the time being 🙇
